# Module 4 Lab 2: Textbook Problem 3.1

**CS 82A - Introduction to Data Science**

**Ramisha Tasfia**

**GitHub:** https://github.com/ramishatasfia5156-commits/cs82a-coursework/blob/main/module4/module4_lab2.ipynb

## Load the Data

In [1]:
import pandas as pd

df = pd.read_csv('allbacks.csv', index_col='rownames')
print('Shape:', df.shape)
df

Shape: (15, 4)


,volume,area,weight,cover
rownames,,,,
1,885,382,800,hb
2,1016,468,950,hb
3,1125,387,1050,hb
4,239,371,350,hb
5,701,371,750,hb
6,641,367,600,hb
7,1228,396,1075,hb
8,412,0,250,pb
9,953,0,700,pb


Weight is in grams, volume in cubic centimeters. `area` is 0 for every paperback in this file (it differs from the textbook table), but area is not used in this problem.

## Median Cover Type (Textbook Part A)

In [2]:
# A median works on ordered categories: sort, then take the middle value
cover_order = pd.CategoricalDtype(['hb', 'pb'], ordered=True)
covers = df['cover'].astype(cover_order).sort_values().reset_index(drop=True)

print(df['cover'].value_counts())
print()
middle = (len(covers) - 1) // 2          # 15 books -> position 8 of 15 (index 7)
print('Middle position:', middle + 1, 'of', len(covers))
print('Median cover:', covers[middle])

cover
pb    8
hb    7
Name: count, dtype: int64

Middle position: 8 of 15
Median cover: pb


**Analysis:** There were 7 hardback copies that come before the 8 paperback copies; therefore, the 8th book out of 15 would be one of the paperback copies. Therefore, the median cover type in this sample is paperback. If we reversed the order and put the paperbacks first, the answer would still be the same because paperbacks make up more than half of the books.

## Mean Weight (Textbook Part B)

In [3]:
print(f"Mean weight, all books: {df['weight'].mean():.2f} g")
print()
by_cover = df.groupby('cover').agg(books=('weight', 'size'),
                                   mean_weight_g=('weight', 'mean'),
                                   mean_volume_cm3=('volume', 'mean')).round(2)
by_cover

Mean weight, all books: 706.67 g



,books,mean_weight_g,mean_volume_cm3
cover,,,
hb,7,796.43,833.57
pb,8,628.12,855.50


In [4]:
diff = by_cover.loc['hb', 'mean_weight_g'] - by_cover.loc['pb', 'mean_weight_g']
print(f'Hardbacks are {diff:.2f} g heavier on average ({diff / by_cover.loc["pb", "mean_weight_g"]:.0%} heavier than paperbacks)')

Hardbacks are 168.31 g heavier on average (27% heavier than paperbacks)


**All Books:** We can estimate that the average weight for a book in this sample is approximately 707 grams.

**By Cover Type:** The average weights for hardback books and paperback books were respectively around 796 g and 628 g. Therefore, there is an approximate difference of 168 grams (27%) between the two cover types, with hardbacks being heavier. Although both have nearly equal average volumes (hardbacks = 834 cm³; paperbacks = 856 cm³), the additional weight of the hardbacks comes from the cover type rather than the hardbacks being larger books. The library may find paperbacks beneficial because their lighter weight would make handling them and stocking shelves easier. However, using only 7 hardbacks and 8 paperbacks as a sample is too limited to base a long-term policy on.

## Variance in Book Volume (Textbook Part C)

In [5]:
print(f"Variance of volume, all books: {df['volume'].var():,.2f} (cm³)²")
print()
df.groupby('cover')['volume'].agg(['mean', 'var', 'std']).round(2)

Variance of volume, all books: 115,518.35 (cm³)²



,mean,var,std
cover,,,
hb,833.57,114100.62,337.79
pb,855.50,132979.71,364.66


The standard deviation for book volume is approximately 340 cm³ overall. The spread is also similar for both cover types; therefore, neither hardbacks nor paperbacks have a narrower range of book sizes. The variability in book volume between the two cover types is very similar.

## Recommendation

The library should select paperback. This is the median cover type in the current sample as well as the more common one. Additionally, even though paperbacks and hardbacks are nearly the same in volume, a paperback weighs about 168 g less than a hardback. The range of book sizes is also about the same for both cover types.

## By-Hand Check: Mean Weight of All Books

Weights: 800, 950, 1050, 350, 750, 600, 1075, 250, 700, 650, 975, 350, 950, 425, 725

Sum = 10,600 g

Mean = 10,600 / 15 = **706.67 g**

In [6]:
weights = [800, 950, 1050, 350, 750, 600, 1075, 250, 700, 650, 975, 350, 950, 425, 725]
print('Sum: ', sum(weights))
print('Mean:', round(sum(weights) / len(weights), 2))
print('Matches pandas:', round(sum(weights) / len(weights), 2) == round(df['weight'].mean(), 2))

Sum:  10600
Mean: 706.67
Matches pandas: True


## AI Disclosure

**AI Disclosure:** Claude (Anthropic) was used to help explain concepts and draft initial code for this lab. All code was reviewed, retyped, and understood by me. I can explain how a median works on ordered categories like cover type, how groupby compares the mean weight of hardbacks and paperbacks, what the variance in book volume shows, and how those results support the cover recommendation.

**Key prompt:** "Help me with Module 4 Lab 2: Textbook Problem 3.1 using allbacks.csv."